File: SBOLCanvasBackend/src/data/GlyphInfo.java
package data;

public class GlyphInfo extends Info {

	private String partType;
	private String[] otherTypes;
	private String partRole;
	private String[] otherRoles;
	private String partRefine;
	private String name;
	private String description;
	private String version;
	private String sequence;
	private String sequenceURI;
	private CanvasAnnotation[] annotations;
	private String[] derivedFroms;
	private String[] generatedBys;

	public String getPartType() {
		return partType;
	}

	public void setPartType(String partType) {
		this.partType = partType;
	}

	public String[] getOtherTypes() {
		return otherTypes;
	}
	
	public void setOtherTypes(String[] otherTypes) {
		this.otherTypes = otherTypes;
	}
	
	public String getPartRole() {
		return partRole;
	}

	public void setPartRole(String partRole) {
		this.partRole = partRole;
	}

	public String[] getOtherRoles() {
		return otherRoles;
	}
	
	public void setOtherRoles(String[] otherRoles) {
		this.otherRoles = otherRoles;
	}
	
	public String getPartRefine() {
		return partRefine;
	}

	public void setPartRefine(String partRefine) {
		this.partRefine = partRefine;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getDescription() {
		return description;
	}

	public void setDescription(String description) {
		this.description = description;
	}

	public String getVersion() {
		return version;
	}

	public void setVersion(String version) {
		this.version = version;
	}

	public String getSequence() {
		return this.sequence;
	}

	public void setSequence(String sequence) {
		this.sequence = sequence;
	}
	
	public String getSequenceURI() {
		return sequenceURI;
	}

	public void setSequenceURI(String sequenceURI) {
		this.sequenceURI = sequenceURI;
	}
	
	public CanvasAnnotation[] getAnnotations() {
		return annotations;
	}

	public void setAnnotations(CanvasAnnotation[] annotations) {
		this.annotations = annotations;
	}
	
	public String[] getDerivedFroms() {
		return derivedFroms;
	}

	public void setDerivedFroms(String[] derivedFroms) {
		this.derivedFroms = derivedFroms;
	}

	public String[] getGeneratedBys() {
		return generatedBys;
	}

	public void setGeneratedBys(String[] generatedBys) {
		this.generatedBys = generatedBys;
	}
	
	public String getFullURI() {
		String fullURI = this.uriPrefix + '/' + this.displayID;
		if(this.version != null && this.version.length() > 0) {
			fullURI += '/' + this.version;
		}
		return fullURI;
	}
}

File: frontend/src/app/glyphInfo.ts
import { ParsedEventType } from '@angular/compiler';
import { CanvasAnnotation } from './canvasAnnotation';
import { Info } from './info';
import { environment } from 'src/environments/environment';

export class GlyphInfo extends Info{
  // Remember that when you change this you need to change the encode function in graph service
  static counter: number = 0;
  partType: string;
  otherTypes: string[];
  partRole: string;
  otherRoles: string[];
  partRefine: string;
  name: string;
  description: string;
  version: string;
  sequence: string;
  sequenceURI: string;
  annotations: CanvasAnnotation[];
  derivedFroms: string[];
  generatedBys: string[];

  constructor(partType?: string) {
    super();
    this.displayID = 'id' + (GlyphInfo.counter++);
    if (partType) {
      this.partType = partType;
    } else {
      this.partType = 'DNA region';
    }
  }

  makeCopy() {
    const copy: GlyphInfo = new GlyphInfo();
    copy.partType = this.partType;
    copy.otherTypes = this.otherTypes ? this.otherTypes.slice() : null;
    copy.partRole = this.partRole;
    copy.otherRoles = this.otherRoles ? this.otherRoles.slice() : null;
    copy.partRefine = this.partRefine;
    copy.displayID = this.displayID;
    copy.name = this.name;
    copy.description = this.description;
    copy.version = this.version;
    copy.sequence = this.sequence;
    copy.sequenceURI = this.sequenceURI;
    copy.uriPrefix = this.uriPrefix;
    copy.annotations = this.annotations ? this.annotations.slice() : null;
    copy.derivedFroms = this.derivedFroms ? this.derivedFroms.slice() : null;
    copy.generatedBys = this.generatedBys ? this.generatedBys.slice() : null;
    return copy;
  }

  copyDataFrom(other: GlyphInfo) {
    this.partType = other.partType;
    this.otherTypes = other.otherTypes ? other.otherTypes.slice() : null;
    this.partRole = other.partRole;
    this.otherRoles = other.otherRoles ? other.otherRoles.slice() : null;
    this.partRefine = other.partRefine;
    this.displayID = other.displayID;
    this.name = other.name;
    this.description = other.description;
    this.version = other.version;
    this.sequence = other.sequence;
    this.sequenceURI = other.sequenceURI;
    this.uriPrefix = other.uriPrefix;
    this.annotations = other.annotations ? other.annotations.slice() : null;
    this.derivedFroms = other.derivedFroms ? other.derivedFroms.slice() : null;
    this.generatedBys = other.generatedBys ? other.generatedBys.slice() : null;
  }

  getFullURI(): string {
    let fullURI = this.uriPrefix + '/' + this.displayID;
    if (this.version && this.version.length > 0) {
      fullURI += '/' + this.version;
    }
    return fullURI;
  }

  encode(enc: any) {
    let node = enc.document.createElement('GlyphInfo');
    if (this.partType)
      node.setAttribute("partType", this.partType);
    if (this.otherTypes) {
      let otherTypesNode = enc.encode(this.otherTypes);
      otherTypesNode.setAttribute("as", "otherTypes");
      node.appendChild(otherTypesNode);
    }
    if (this.partRole)
      node.setAttribute("partRole", this.partRole);
    if (this.otherRoles) {
      let otherRolesNode = enc.encode(this.otherRoles);
      otherRolesNode.setAttribute("as", "otherRoles");
      node.appendChild(otherRolesNode);
    }
    if (this.partRefine)
      node.setAttribute("partRefine", this.partRefine);
    if (this.displayID)
      node.setAttribute("displayID", this.displayID);
    if (this.name && this.name.length > 0)
      node.setAttribute("name", this.name);
    if (this.description && this.description.length > 0)
      node.setAttribute("description", this.description);
    if (this.version && this.version.length > 0)
      node.setAttribute("version", this.version);
    if (this.sequence && this.sequence.length > 0)
      node.setAttribute("sequence", this.sequence);
    if(this.sequenceURI && this.sequenceURI.length > 0)
      node.setAttribute("sequenceURI", this.sequenceURI);
    if (this.uriPrefix)
      node.setAttribute("uriPrefix", this.uriPrefix);

    if (this.annotations) {
      let annotationsNode = enc.encode(this.annotations);
      annotationsNode.setAttribute("as", "annotations");
      node.appendChild(annotationsNode);
    }
    if(this.derivedFroms){
      let derivedFromsNode = enc.encode(this.derivedFroms);
      derivedFromsNode.setAttribute("as", "derivedFroms");
      node.appendChild(derivedFromsNode);
    }
    if(this.generatedBys){
      let generatedBysNode = enc.encode(this.generatedBys);
      generatedBysNode.setAttribute("as", "generatedBys");
      node.appendChild(generatedBysNode);
    }

    return node;
  }
}
